// themeweaver/lib/test_themes/explicitValuesTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.theme = void 0;

var _index = require("../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var theme = (0, _index.withThemeweaver)({
  colors: {
    primary: '#7789C8',
    secondary: '#C6D8FF',
    action: ['#F6E8E8', '#E5707A'],
    lightText: '#979797',
    mainText: '#444649',
    secondaryText: '#E5707A',
    lightBackground: '#F8F8F8',
    disabledBackground: '#E2E2E2',
    white: '#ffffff'
  },
  fonts: {
    poppins: "'Poppins', sans-serif",
    roboto: "'Roboto', sans-serif",
    openSans: "'Open Sans', sans-serif"
  },
  fontSizes: [10, 12, 14, 18, 24, 32],
  fontWeights: {
    regular: '400',
    bold: '700'
  },
  lineHeights: ['normal', 2, 3],
  letterSpacings: ['normal', '.2rem', 1],
  sizes: [0, '100%', 80],
  space: [0, 4, 8, 16, 20, 32, 64],
  borders: ['dashed red', 'thick double #32a1ce'],
  borderWidths: ['3px 5px 3px 5px', '3px'],
  borderStyles: ['none', 'solid'],
  radii: [0, '5px', '20px'],
  shadows: ['1px 1px 2px pink', '5px 5px #558ABB'],
  zIndices: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '999999'],
  transitions: {
    down: 'top 500ms ease-in-out',
    left: 'left 500ms ease',
    right: 'right 500ms ease',
    colorBig: 'color 500ms ease',
    padding: 'padding 500ms ease'
  },
  breakpoints: ['0', '40em', '52em', '64em', '80em']
}, {
  // beginning of themeweaver asignments *****************
  document: {
    paddingLeft: (0, _index.value)('5px'),
    paddingRight: 2
  },
  content: {
    paddingLeft: 2
  },
  button: _objectSpread({}, (0, _index.variant)('nav', {
    height: (0, _index.value)('50px'),
    width: (0, _index.value)('50px'),
    color: 'action[1]'
  })),
  nav: _objectSpread(_objectSpread({}, (0, _index.variant)('main', {
    backgroundColor: (0, _index.value)('#C6D8FF'),
    border: (0, _index.value)('dashed blue'),
    borderTop: (0, _index.value)('solid blue'),
    borderRight: (0, _index.value)('solid'),
    borderBottom: (0, _index.value)('dashed'),
    borderLeft: (0, _index.value)('solid red'),
    borderColor: (0, _index.value)('#F8F8F8'),
    borderRadius: (0, _index.value)('5px'),
    borderStyle: (0, _index.value)('dashed solid'),
    borderWidth: (0, _index.value)('5px'),
    boxShadow: (0, _index.value)('1px 1px 1px 1px blue'),
    color: (0, _index.value)('red'),
    fontFamily: (0, _index.value)("'Poppins', sans-serif"),
    fontSize: (0, _index.value)('18px'),
    fontWeight: (0, _index.value)('500'),
    height: (0, _index.value)('80px'),
    lineHeight: (0, _index.value)('3'),
    letterSpacing: (0, _index.value)('2.5'),
    margin: (0, _index.value)('50px'),
    marginTop: (0, _index.value)('51px'),
    marginRight: (0, _index.value)('52px'),
    marginBottom: (0, _index.value)('53px'),
    marginLeft: (0, _index.value)('54px'),
    maxHeight: (0, _index.value)('55px'),
    maxWidth: (0, _index.value)('56px'),
    minHeight: (0, _index.value)('57px'),
    minWidth: (0, _index.value)('58px'),
    padding: (0, _index.value)('59px'),
    paddingTop: (0, _index.value)('60px'),
    paddingRight: (0, _index.value)('61px'),
    paddingBottom: (0, _index.value)('62px'),
    paddingLeft: (0, _index.value)('63px'),
    textShadow: (0, _index.value)('1px 2px 1px 2px blue'),
    transition: (0, _index.value)('top 500ms ease-in-out'),
    width: (0, _index.value)('70px'),
    zIndex: (0, _index.value)('1')
  })), (0, _index.variant)('hover', {
    color: 'action[0]',
    backgroundColor: 'action[0]',
    fontFamily: 'poppins',
    fontSize: 3,
    fontWeight: 'bold',
    padding: 6
  }))
});
exports.theme = theme;
var _default = theme;
exports["default"] = _default;
