// themeweaver/lib/test_themes/theme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.theme = void 0;

var _index = require("../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var theme = (0, _index.withThemeweaver)({
  colors: {
    primary: '#7789C8',
    secondary: '#C6D8FF',
    action: ['#F6E8E8', '#E5707A'],
    lightText: '#979797',
    mainText: '#444649',
    secondaryText: '#E5707A',
    lightBackground: '#F8F8F8',
    disabledBackground: '#E2E2E2',
    white: '#ffffff'
  },
  fonts: {
    poppins: "'Poppins', sans-serif",
    roboto: "'Roboto', sans-serif",
    openSans: "'Open Sans', sans-serif"
  },
  fontSizes: [10, 12, 14, 18, 24, 32],
  fontWeights: {
    regular: 400,
    bold: 700
  },
  sizes: [0, '100%', 80],
  space: [0, 4, 8, 16, 20, 32, 64],
  transitions: {
    down: 'top 500ms ease-in-out',
    left: 'left 500ms ease',
    right: 'right 500ms ease',
    colorBig: 'color 500ms ease',
    padding: 'padding 500ms ease'
  },
  breakpoints: ['40em', '52em', '64em', '80em']
}, {
  // beginning of themeweaver asignments *****************
  document: {
    paddingLeft: 2,
    paddingRight: 2
  },
  content: {
    paddingLeft: 2
  },
  button: _objectSpread({}, (0, _index.variant)('nav', {
    height: (0, _index.value)('50px'),
    width: (0, _index.value)('50px'),
    color: 'action[1]'
  })),
  nav: _objectSpread(_objectSpread({}, (0, _index.variant)('main', {
    color: ['action[1]', 'lightText'],
    backgroundColor: 'secondary',
    fontFamily: 'poppins',
    fontSize: 3,
    fontWeight: 'bold',
    marginLeft: [0, 2],
    marginRight: [0, 2],
    marginTop: 0,
    marginBottom: 0,
    padding: 4,
    paddingTop: [0, 2],
    paddingLeft: 0,
    transition: {
      top: 'down',
      left: 'left'
    },
    minHeight: (0, _index.value)('80px')
  })), (0, _index.variant)('hover', {
    color: 'action[0]',
    backgroundColor: 'action[0]',
    fontFamily: 'poppins',
    fontSize: 3,
    fontWeight: 'bold',
    padding: 6
  }))
});
exports.theme = theme;
var _default = theme;
exports["default"] = _default;
